lib/facter/mysql_version.rb

```diff
@@ -7,3 +7,11 @@
     mysql_ver.match(%r{\d+\.\d+\.\d+})[0] if mysql_ver
   end
 end
+
+Facter.add('mysql_version') do
```

Comment:
now we've two blocks for the same fact. Doesn't it make sense to unify it? What if mysql and mariadb is installed on the same server?

Reply:
This documentation feels weird.  Inspecting various facts in the puppetlabs organization repositories, I don't see usage of confine with different prerequisites for the same facts, but a single fact with confinement that checks multiple conditions, e.g. https://github.com/puppetlabs/puppetlabs-java/blob/6319799bb310f824b18d9077905c46409535c977/lib/facter/java_default_home.rb#L16

It feels more natural for me to have a single `Facter.add('mysql_version')`, where the setcode logic attempts to get the expected value and return nil if none was found.  A confinement does not even provide a short-circuit in this case, so we can probably go without it, i.e.

```ruby
Facter.add('mysql_version') do
  setcode do
    if Facter::Core::Execution.which('mysql')
      mysql_ver = Facter::Core::Execution.execute('mysql --version')
      mysql_ver.match(%r{\d+\.\d+\.\d+})[0] if mysql_ver
    elsif Facter::Core::Execution.which('mariadb')
      mysql_ver = Facter::Core::Execution.execute('mariadb --version')
      mysql_ver.match(%r{\d+\.\d+\.\d+})[0] if mysql_ver
    end
  end
end
```

So merging may make sense IMHO.


But beyond this consideration (and speaking as someone who does not use mysql nor mariadb), if mariadb and mysql are diverging one from the other and could be installed side by side on a single system, maybe it make sense to "fork" the puppetlabs-mysql module (puppet(labs)?-mariadb) just like mariadb forked mysql so that each module can track what the software they configure implement and provide a consistent interface and not some weak abstraction that may or may not work depending on the package you manage with the module?

Reply:
Facters reworked and correctly working.
Thank you @smortex , code is much nicer !